staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -287,3 +290,29 @@ func (s *Validator) validateArray(ctx context.Context, fldPath *field.Path, sts
 
 	return errs, remainingBudget
 }
+
+// OpenAPIv3 type/maxLength/maxItems/MaxProperties/required/wrong type field validation failures are viewed as blocking err for CEL validation
+func hasBlockingErr(errs field.ErrorList, fldPath *field.Path) bool {
+	for _, err := range errs {
+		if strings.HasPrefix(err.Field, fldPath.String()) &&
```

Comment:
if I remember correctly, fldPath.String() is pretty expensive, allocation-wise, and we're calling hasBlockingErr at every ancestor of every level that contains CEL rules

come to think of it, err.Error() is also doing fmt.Sprintf() on every call, which this function is calling 6 times

I think it would be better to have something like `isCelBlockingErr(err *field.Error) bool` that we can test in isolation, and filter any existing errors down to just blocking errors once at the top level, so we don't have to do these checks over and over at each level

Reply:
I'm also on the fence about trying to run CEL expressions in unrelated parts of the object when there are fundamental errors like type errors or missing fields, or maxLength/maxItems/maxProperties errors. it would certainly be simpler to only run dynamic validation on well-typed objects that specify required fields and don't exceed static size limits. @jpbetz, do you think it's important to do partial validation like this?

Reply:
I have updated to use func isCelBlockingErr on err type for blocking error checking. Only run dynamic validation on well-typed objects sounds reasonable to me as well. Let me know if it's ok to go with this.